### ISSUE TITLE
No MKL Compatible Conda installation for PyTorch 1.5

### ISSUE THREAD

**Issue by MichaelLutter** [pytorch/pytorch#38412]
## 🚀 Feature
Right now having the conda nomkl package installed is a comfortable way to ensure that numpy is linking to openblas instead of Intel mkl, which is (in my experience) slower for AMD processors. If the no mkl package is installed, the conda pytorch installation resolves to an older pytorch version.

I.e. the command 
`conda install pytorch torchvision cudatoolkit=10.2 -c pytorch`  

installs WITHOUT the nomkl package
`pytorch            pytorch/linux-64::pytorch-1.5.0-py3.7_cuda10.2.89_cudnn7.6.5_0`

and WITH the nomkl package
`pytorch            pytorch/linux-64::pytorch-1.0.0-py3.7_cuda9.0.176_cudnn7.4.1_1`

It would be great to have a conda compatible installation with nomkl, as AMD processors have gotten good in terms of # of fast cores and value. So if it would not require to much effort it would be great to have. And also a bit of documentation on how to ensure linking to openblas instead of mkl would be great as AMD processor have become more widespread in recent years.

## Alternatives
To get pytorch 1.5 with nomkl, one can simply install pytorch 1.5 via pip.

**Comment by peterjc123** [comment]
It seems that MKL is also usable with AMD CPUs, but you'll need to set `MKL_DEBUG_CPU_TYPE=5` as the environment variable.

Reference:
https://www.extremetech.com/computing/302650-how-to-bypass-matlab-cripple-amd-ryzen-threadripper-cpus

**Comment by rkaplan** [comment]
Thank you for creating this issue. I spent a while tracking down why I couldn't bump the pytorch version from 1.4 to 1.5 in my conda environment file, before finally discovering this is the problem. The symptom is, when running `conda env update -f myenv.yaml` with an environment yaml that contains both `nomkl` and `pytorch>=1.5`:

```
Found conflicts! Looking for incompatible packages.
This can take several minutes.  Press CTRL-C to abort.
```

Moving the pytorch installation to pip, as suggested by the issue author, resolved the issue for me. My env file now looks something like:

```
name: myenv
channels:
  - conda-forge
dependencies:
  - python=3.7
  - pip
  - ...
  - pip:
    - torch>=1.5
    - torchvision
```

It would be great to fix this and make conda pytorch 1.5 play nicely with nomkl.

**Comment by hz838** [comment]
The pytorch 1.5 installed via pip seems to be based on MKL too (Windows OS) while the numpy installed via pip is based on openBLAS.

Will pytorch be slower on AMD CPUs if I use CPU to train models?

**Comment by hz838** [comment]
> 
> 
> It seems that MKL is also usable with AMD CPUs, but you'll need to set `MKL_DEBUG_CPU_TYPE=5` as the environment variable.
> 
> Reference:
> https://www.extremetech.com/computing/302650-how-to-bypass-matlab-cripple-amd-ryzen-threadripper-cpus

It is said that this trick has been banned in MKL 2020. I am not sure because I am only considering to purchase an AMD PC and I am using Intel CPUs currently.

**Comment by peterjc123** [comment]
I'm not sure whether we will build binaries with openblas or not but I think we should make sure the build passes with openblas.

**Comment by matrs** [comment]
> It seems that MKL is also usable with AMD CPUs, but you'll need to set `MKL_DEBUG_CPU_TYPE=5` as the environment variable.
> 
> Reference:
> https://www.extremetech.com/computing/302650-how-to-bypass-matlab-cripple-amd-ryzen-threadripper-cpus

It's usable, it always has been, but it's slower than `openblas` in linux as the OP says:

https://www.pugetsystems.com/labs/hpc/AMD-Ryzen-3900X-vs-Intel-Xeon-2175W-Python-numpy---MKL-vs-OpenBLAS-1560/

I arrived here because I wanted to create a new env using `conda` with `pytorch` (pytorch channel ) and `nomkl` and I couldn't. As the OP says, it would be good to not require `MKL`, AMD has sold a lot of CPUs the last years and it will continue to do so in the near future, I think that's a good reason to give an alternative to MKL.

**Comment by ekerazha** [comment]
@matrs 
The `MKL_DEBUG_CPU_TYPE=5` trick has been disabled since version 2020.1

**Comment by ekerazha** [comment]
Related issue: https://github.com/pytorch/pytorch/issues/37746

**Comment by matrs** [comment]
> @matrs
> The `MKL_DEBUG_CPU_TYPE=5` trick has been disabled since MKL version 2020.1 and it doesn't work anymore.
> 
> Building pytorch packages with MKL instead of OpenBLAS makes them unsuitable for AMD cpus.

You don't need that trick for AMD CPUs to work with MKL. That trick is to make it **faster with AMD cpus, not to work with AMD**. MKL works with AMD without any trick, it's just slower.

**Comment by hz838** [comment]
> 
> 
> > @matrs
> > The `MKL_DEBUG_CPU_TYPE=5` trick has been disabled since MKL version 2020.1 and it doesn't work anymore.
> > Building pytorch packages with MKL instead of OpenBLAS makes them unsuitable for AMD cpus.
> 
> You don't need that trick for AMD CPUs to work with MKL. That trick is to make it **faster with AMD cpus, not to work with AMD**. MKL works with AMD without any trick, it's just slower.

It seems that this trick has been banned and it's useless now. 

I tried it on my AMD 4800H laptop with Windows 10, Python 3 and Pytorch 1.5, it seems that it has no impact on the speed of my neural networks.

**Comment by ekerazha** [comment]
> You don't need that trick for AMD CPUs to work with MKL. That trick is to make it **faster with AMD cpus, not to work with AMD**. MKL works with AMD without any trick, it's just slower.

Yeah it's like running it on a Pentium III "Katmai" from year 1999, but it works.

**Comment by kienerj** [comment]
> 
> 
> @matrs
> The `MKL_DEBUG_CPU_TYPE=5` trick has been disabled since MKL version 2020.1 and it doesn't work anymore.
> 
> Building pytorch packages with MKL instead of OpenBLAS makes them unsuitable for AMD cpus.

Is this actually true? Where does this info/rumor come from? I just played around on my system (AMD Ryzen 3900x) and I see no difference wether I use openblas version of of numpy or the mkl version. In fact the mkl one is a bit faster. conda states I have MKL 2020.1 installed with numpy 1.19.1. Removing the `MKL_DEBUG_CPU_TYPE=5` flag doesn't have an effect.

In fact from my side it looks like the flag isn't needed at all to get good performance from Ryzen with below code:

```
n = 20000
A = np.random.randn(n,n).astype('float64')
B = np.random.randn(n,n).astype('float64')
```

```
%%timeit
np.linalg.norm(A@B)
```

I ran that previously and the flag had a huge effect and openblas was a lot faster than mkl without the flag. Hence my conclusion mkl 2020.1 actually fixes the issue for Ryzen CPUs by default and hence the flag is not needed anymore / has no effect.

**Comment by ekerazha** [comment]
@kienerj 
It has been confirmed by multiple users that Intel MKL 2020.1 performances are poor on AMD and the MKL_DEBUG_CPU_TYPE workaround doesn't work anymore, just read the latest comments here:
- https://www.reddit.com/r/matlab/comments/dxn38s/howto_force_matlab_to_use_a_fast_codepath_on_amd/?sort=new
- https://www.reddit.com/r/MachineLearning/comments/f2pbvz/discussion_workaround_for_mkl_on_amd/?sort=new

Moreover, thinking that Intel would enable the fast code path on AMD CPUs is hilarious.

**Comment by kienerj** [comment]
@ekerazha 

I agree that I'm literally shocked that on my AMD Ryzen CPU with Anaconda Python which uses MKL 2020.1 I get by default the fast performance.

The  MKL_DEBUG_CPU_TYPE settings has no effect because it's fast by default. I played around with this end of last year and was able to 100% repeat [these results](https://www.pugetsystems.com/labs/hpc/AMD-Ryzen-3900X-vs-Intel-Xeon-2175W-Python-numpy---MKL-vs-OpenBLAS-1560/). Yes, MKL was slow end of 2019 on AMD CPU without the setting.  I also have a 3900x and my numbers matched the results pugestsystems got. 

I retested yesterday and MKL 2020.1 from Anaconda simply is by default fast, faster than OpenBLAS in all tests and in some tests like SVD mkl is about 3times faster than OpenBLAS (using numpy to test that). I played around for pretty long with several reboots and checks. It's true. Get access to a ryzen/threadripper/epyc machine and try yourself. 

I also have access to intel machine (my work laptop) with 6-cores and in the pugestsystems dot product test it takes pretty much exactly double the time (a bit more) compared to my 3900x with 12-cores. (I'm aware of core differences but it's still shows that indeed MKL 2020.1 works fine on AMD CPU).

About your links:

- First one shows that Matlab was fixed. How exactly do you think it was fixed? MKL for intel, OpenBLAS for AMD? Maybe. But wouldn't it be much easier to simply force your supplier (Intel) to fix their library? Yeah not proof but I actually suspect we can thank MathWorks for this. Yeah, pure conjecture but conjecture based on rational arguments.

- Second links the [second top level comment](https://old.reddit.com/r/MachineLearning/comments/f2pbvz/discussion_workaround_for_mkl_on_amd/fvgupoi/) literally has the same observation that I'm having, MKL 2020.1 being faster than OpenBLAS and nedflanders1976 second comment to this reply is simply wrong. MKL was dog slow end of last year, see pugetssystems links above. It was just a bit slow, it was unusable slow.

Hell really froze over it seems. Don't get me wrong. I'm shocked as well hence I did a lot of testing and checks.

**Comment by hz838** [comment]
> 
> 
> @ekerazha
> 
> I agree that I'm literally shocked that on my AMD Ryzen CPU with Anaconda Python which uses MKL 2020.1 I get by default the fast performance.
> 
> The MKL_DEBUG_CPU_TYPE settings has no effect because it's fast by default. I played around with this end of last year and was able to 100% repeat [these results](https://www.pugetsystems.com/labs/hpc/AMD-Ryzen-3900X-vs-Intel-Xeon-2175W-Python-numpy---MKL-vs-OpenBLAS-1560/). Yes, MKL was slow end of 2019 on AMD CPU without the setting. I also have a 3900x and my numbers matched the results pugestsystems got.
> 
> I retested yesterday and MKL 2020.1 from Anaconda simply is by default fast, faster than OpenBLAS in all tests and in some tests like SVD mkl is about 3times faster than OpenBLAS (using numpy to test that). I played around for pretty long with several reboots and checks. It's true. Get access to a ryzen/threadripper/epyc machine and try yourself.
> 
> I also have access to intel machine (my work laptop) with 6-cores and in the pugestsystems dot product test it takes pretty much exactly double the time (a bit more) compared to my 3900x with 12-cores. (I'm aware of core differences but it's still shows that indeed MKL 2020.1 works fine on AMD CPU).
> 
> About your links:
> 
>     * First one shows that Matlab was fixed. How exactly do you think it was fixed? MKL for intel, OpenBLAS for AMD? Maybe. But wouldn't it be much easier to simply force your supplier (Intel) to fix their library? Yeah not proof but I actually suspect we can thank MathWorks for this. Yeah, pure conjecture but conjecture based on rational arguments.
> 
>     * Second links the [second top level comment](https://old.reddit.com/r/MachineLearning/comments/f2pbvz/discussion_workaround_for_mkl_on_amd/fvgupoi/) literally has the same observation that I'm having, MKL 2020.1 being faster than OpenBLAS and nedflanders1976 second comment to this reply is simply wrong. MKL was dog slow end of last year, see pugetssystems links above. It was just a bit slow, it was unusable slow.
> 
> 
> Hell really froze over it seems. Don't get me wrong. I'm shocked as well hence I did a lot of testing and checks.

I'm not en experienced user. Due to COVID-19 I'm using my own computers in recent months and I'm using an AMD 4800H laptop and an Intel i7-10700F desktop. Both are running Windows 10.

When I am running the graph data attack algorithm, the 10700F desktop is around 2.5 times faster than the 4800H laptop. On the other hand when I encode videos with x264, the Intel dekstop is only around 20% faster. The pytorch version I am running is installed through pip and I have no openBLAS version.

**Comment by ekerazha** [comment]
@kienerj 
I wrote "read the latest comments", not the articles themselves, which contain obsolete information.

Matlab "fixed" it by using the MKL_DEBUG_CPU_TYPE trick, which doesn't work anymore since MKL 2020.1. Matlab still uses MKL 2019.x. Actually, comment from the "nedflanders1976" was right.

**Comment by kienerj** [comment]
> I'm not en experienced user. Due to COVID-19 I'm using my own computers in recent months and I'm using an AMD 4800H laptop and an Intel i7-10700F desktop. Both are running Windows 10.
> 
> When I am running the graph data attack algorithm, the 10700F desktop is around 2.5 times faster than the 4800H laptop

Have you ensured you are using same version of everything on both machines and it's MKL 2020.1?

I'm not an experienced PyTorch users so does graph data attack algorithm run only on CPU? Or does the desktop possibly run some parts on the GPU? The F in i7-10700F means no integrated GPU meaning you have a dedicated graphics card in that machine and then a 2.5 speed up woudl actually not be a lot at all.



Even if you ensure above points, the result still is explainable as the desktop CPU will have much better cooling and will run at much higher clock speed during a long running test. Don't trust the Intel TDP numbers, most motherboards will run Intel CPUs at much higher power use, Your i7-10700F is mostly likley running at around 4 ghz for most of the test while the laptop most likley will run at base clock of 2.9 GHZ or throttle even lower depending on the actual cooling (laptop model) and your room temperature. On top of that the 4800H has half the L3 cache compared to desktop Ryzen. That could also play a role. 

I'm being skeptical because I'm seeing the results on my machine and there is no doubt that my claim is true for Anaconda MKL 2020.1 on Windows 10. Are you on windows or Linux? Maybe it's different for Linux?

**Comment by kienerj** [comment]
@ekerazha 

On which AMD CPU are you testing this?

I actually own one and can test and there is 0 doubt that for my setup my claim is true.
Setup= numpy from anaconda channel with MKL 2020.0.1 on windows 10 and a 3900x without the fix.

- with mkl all the tests are similar or faster, sometimes a lot, than compared to OpenBLAS on the 3900x machine
  - svd, cholesky and eig are much slower with OpenBLAS
- Also the tests are a lot faster than they where end of 2019 without the fix
- 3900x is in some tests (matrix /vector multiplication) 2x times faster than 6-core i7-8850H (makes sense)
- in svd, cholesky and eig the i7-8850H is about the same as the 3900x but are these fully multi-threaded?
  - for these tests you might have a point but cache and memory optimization could play a big role. Making AMD CPU use AVX/AVX2 doesn't mean they profit from Intel specific cache optimizations. 

Conclusion:
- MKL 2020.0.1 Product Build 20200208 is just as fast as older MKL version with MKL_DEBUG_CPU_TYPE fix
- MKL 2020.0.1 is faster than OpenBLAS, in some test a lot faster

Hence even if MKL hinders AMD CPU in svd, eig and cholesky, it's still faster than using OpenBLAS.

**Comment by ekerazha** [comment]
@kienerj 
It's likely that you are not correctly setting/unsetting the environment variable when performing the tests.

**Comment by kienerj** [comment]
> 
> 
> @kienerj
> It's likely that you are not correctly setting/unsetting the environment variable when performing the tests.

I'm sure it's not set because displaying it's value is part of the "benchmark" I'm running. 

But more importantly your claim is that the variable doesn't work anymore so having it set should not matter anyway => you always get slow performance. And this is simply not the case.  The opposite is the case, you always get the fast code path.*

This is my "benchmark script":

![mkl benchmark](https://user-images.githubusercontent.com/1567309/90612915-f9654200-e208-11ea-84e0-baba8045bb05.png)


* I previously wrote that  svd, cholesky and eig aren't any faster on 3900x vs 6-core intel. Some further testing simply showed they don't scale well at all with core count beyond 2 cores.

**Comment by ekerazha** [comment]
@kienerj 
This is how you should perform the test:
- Link numpy to MKL 2020.0, test with and without the MKL_DEBUG_CPU_TYPE trick. It is faster with the trick.
- Link numpy to MKL 2020.1, test with and without the MKL_DEBUG_CPU_TYPE trick. In both cases, it is slow just like the 2020.0 version without the trick. This has been confirmed by multiple users, as I showed before.

You can use the code you can find here for comparison: https://www.pugetsystems.com/labs/hpc/AMD-Ryzen-3900X-vs-Intel-Xeon-2175W-Python-numpy---MKL-vs-OpenBLAS-1560/

If you get different results, you are doing something wrong.

**Comment by kienerj** [comment]
I can only lead to horse to water. Anyone that actually owns a Ryzen CPU can easily confirm my findings (anaconda python with numpy and mkl 2020.1, windows) and I encourage them to do so. After all you have no reason to believe either of us so testing yourself is the best one can do. See attached the notebook I used to come up with my claims (zipped as plain isn't supported)

[BLAS benchmarking numpy.zip](https://github.com/pytorch/pytorch/files/5098482/BLAS.benchmarking.numpy.zip)


@ekerazha 

That link is straight from my initial comment and yes back in late 2019, this is what happened. The test took around 125s with MKL on Ryzen vs 39s with OpenBLAS. I ran these test back then already and they were back then 100% correct. But with MKL 2020.0.1 it takes 34s without the fix set (fix has no effect as fast by default) and about 90s on a 6-core intel laptop showing that indeed intel has no advantage at all in this test.

On top of that as also said multiple times the MKL version is faster in all tests over OpenBLAS, in some a lot faster. That would simply not be possible if MKL used only SSE.

You can also run the attached test on your system and compare to my numbers, they are in the notebook. Only thing I can possibly see is that the behavior is different depending on OS. I'm on Windows 10.

**Comment by ekerazha** [comment]
@kienerj 
I encourage them too. Meanwhile nobody confirmed what you say and multiple users confirmed the opposite. Moreover, comparing different systems with totally different hardware makes no sense. You have to follow the steps I wrote above, same system, same configuration, different MKL versions with and without the MKL_DEBUG_CPU_TYPE workaround, so numbers are comparable. Science. Less recent versions of OpenBLAS could provide poor performances because of a known bug in CPU detection.

**Comment by huanzhang12** [comment]
I did a test on three AMD Zen based systems (Zen 1, Zen+ and Zen 2) and I can confirm that:

1. The `MKL_DEBUG_CPU_TYPE` flag is removed in latest MKL (I tested 2020.2)
2. On AMD systems, the behavior on MKL 2020.2 is different than MKL 2020.0 (and earlier versions), however **it is still very slow**.

Nevertheless, **we should not rely on MKL's unpredictable behavior on AMD systems**. AMD CPUs are currently performance leaders in the x86 ecosystem and we expect to see more and more AMD systems in HPC settings. For example, the latest Nvidia DGX A100 workstations are based on AMD EPYC 7742 CPUs. **It is necessary and crucial to provide pytorch builds without MKL** (using blis or OpenBlas instead).

**Details:**

I used this simple script (as used [here](https://www.pugetsystems.com/labs/hpc/How-To-Use-MKL-with-AMD-Ryzen-and-Threadripper-CPU-s-Effectively-for-Python-Numpy-And-Other-Applications-1637/)) to measure time:

```python
import numpy as np
import time
n = 20000
A = np.random.randn(n,n).astype('float64')
B = np.random.randn(n,n).astype('float64')
start_time = time.time()
nrm = np.linalg.norm(A@B)
print(" took {} seconds ".format(time.time() - start_time))
print(" norm = ",nrm)
```

The performance measurements are listed in this table (I ran each setting three times and took the average):

|                                       | EPYC 7702 (Zen 2) | EPYC 7601 (Zen) | Ryzen 7 2700X (Zen+) |
|---------------------------------------|-------------------|-----------------|----------------------|
| MKL 2020.0 (MKL_DEBUG_CPU_TYPE=5)     | **10.58**         | **21.60**       | **68.69**            |
| MKL 2020.0 (unset MKL_DEBUG_CPU_TYPE) | 29.85             | 33.86           | 72.58                |
| MKL 2020.2 (MKL_DEBUG_CPU_TYPE=5)     | 26.27             | 24.31           | 72.98                |
| MKL 2020.2 (unset MKL_DEBUG_CPU_TYPE) | 25.49             | 24.24           | 72.76                |

It is obvious that MKL 2020.0 with `MKL_DEBUG_CPU_TYPE=5` is best performing. Especially, on Zen 2 based systems (EPYC 7002 and Ryzen 3000 series), MKL 2020.2 is slow, only slightly better than MKL 2020.0 without `MKL_DEBUG_CPU_TYPE=5`.

As a temporary workaround, you can force install MKL 2020.0 and set `export MKL_DEBUG_CPU_TYPE=5`:

```bash
conda install mkl=2020.0
```

**Comment by kienerj** [comment]
@huanzhang12 

Agree that a intel independent build ready to go is a good thing. I can't repeat your tests as in I'm getting same performance in all cases on my 3900x. Maybe Intel are limiting core-scaling which doesn't affect my 3900x but does your 64-core epyc? No idea. I score at 30s for this test (windows 10!!!)

But it seems we both are right. See [this issue](https://github.com/pytorch/builder/issues/504) from pytorch-builder linking to [this hacker news thread.](https://news.ycombinator.com/item?id=24313442)

**Comment by rkaplan** [comment]
This article came out today with useful information on how to patch MKL 2020.2 to maintain the old `MKL_DEBUG_CPU_TYPE=5` speedup even though that trick no longer works: https://danieldk.eu/Posts/2020-08-31-MKL-Zen.html

tl;dr Compile a shared library that overrides `mkl_serv_intel_cpu_true`.

**Comment by malfet** [comment]
@huanzhang12 can you also please benchmark the performance of MKL 2020.3, which claims to have fixed the perf regressions introduced in MKL 2020.2

**Comment by olejorgenb** [comment]
What is the best workaround here? We use both pytorch in production which force us to a blas based on mkl which appear to slow down certain numpy operations by ~10-300x  (!!) (AMD EPYC 7453)

**Comment by bsergean** [comment]
fyi I did a test with very recent versions of pytorch, and it's still the same, at least for matrix multiplication, MKL fallback to some SSE code. I can get fast perf by recompiling pytorch with blis, but it's cumbersome.

Overriding mkl_serv_intel_cpu_true is a cool trick, I don't know if it still works.
